create_layout.py

```diff
@@ -91,45 +91,120 @@
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a file structure -- 'before' and 'after' -- 
+    return a tuple specifying which files have been added, 
+    which have been removed, which have been modified, 
+    and which have remained unchanged. 
+    Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = []
+  modified_files = []
+  added_files = []
+  removed_files = []
+  for key in before_dict:
+    if key in after_dict:
+      if before_dict[key] == after_dict[key]:
+        # Matching the hashes to check if file was unchanged
+        unchanged_files.append(key)
+      else:
+        modified_files.append(key)
+    else:
+      removed_files.append(key)
+  for key in after_dict:
+    if key not in before_dict:
+      # Looking for new files
+      added_files.append(key)
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+      sorted(removed_files))
+
+def create_material_rules(changes_of_curr_step, links, index):
   """Create generic material rules (3 variants)
 
+  changes_of_curr_step: changes between the current step's materials and products
+      - a return value of the changes_between_snapshots() function when 
+      the current step's materials and products are the values passed to the function
+  links: an ordered list of in-toto link objects
+  index: the index of the current step
+
   * MATCH available materials with products from previous step (links must be an
   ordered list) and
   * ALLOW available materials if it is the first step in the
   list
+  * DELETE removed materials
+
   Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
+  NOTE: Read header docstring for ideas for more complexity.  
+  """
 
   expected_materials = []
+  unchanged_files, modified_files, _, deleted_files = changes_of_curr_step
+  curr_step_materials = unchanged_files + modified_files + deleted_files
```

Comment:
I may be missing something here, but `DELETE` doesn't apply as a material rule, but rather as a product rule for artifacts that are deleted as part of that corresponding step.

```DELETE: indicates that materials matched by the pattern must not appear as products of this step.```

Reply:
I discussed this with Lukas in the previous PR, and he suggests me to put `DELETE` in the materials rules.
https://github.com/in-toto/layout-web-tool/pull/46#discussion_r464405264

I think what he means is that becuase deleted artifacts won't show up in the product queue, we should put `DELET` in the materilas rules. What do you think?

Reply:
Yeah, looks like I'd misunderstood how the delete rule applies in practice. Carry on!